### PR TITLE
[RFC] Add a pull request template

### PR DESCRIPTION
**Pull request by wilzbach**
This is a proposal for experimenting with a pull request template.

More details about this:
- https://github.com/blog/2111-issue-and-pull-request-templates
- https://help.github.com/articles/creating-a-pull-request-template-for-your-repository
- https://github.com/devspace/awesome-github-templates

I wasn't very good with coming up with essential questions or reminders.
- What did I forget?
- What isn't needed?

### CODE REVIEW

File: PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+### Bug fixes:
+
+- [ ] Is the Bugzilla issue referenced in the commit message?
+      (check https://github.com/dlang-bots/dlang-bot#automated-references for help)
+- [ ] Have you included a test for the fixed issue?
+- [ ] Have you targeted `stable` (major, critical and regression fixes)?
+
+### Core changes:
+
+* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
```
Comment (JackStouffer) [inline PULL_REQUEST_TEMPLATE.md]: I would take this one out and reword it in the "Description" section, like so

> Be sure to include a detailed explanation of what your changes do and why you'd like us to include them

File: PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+* [ ] Have you written new tests for your core changes, as applicable?
+* [ ] Have you added a [changelog entry](https://github.com/dlang/phobos/tree/master/changelog)?
```
Comment (JackStouffer) [inline PULL_REQUEST_TEMPLATE.md]: Mention they should only do this for brand new features, breaking changes, or performance increases.

File: PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+### Description
+
+Please describe your pull request.
+
+💔 Thank you!
```
Comment (JackStouffer) [inline PULL_REQUEST_TEMPLATE.md]: I would remove this. This is a template and not a set of instructions after all.

**Comment by wilzbach** [comment]
Thanks a lot for your feedback @JackStouffer!
A couple of additional ideas for checks:

- [ ] I have added DDoc documentation to the new symbols (if applicable)
- [ ] I have checked the coverage output files (e.g. `make -f std/algorithm/iteration.test` and then look for `std/algorithm/iteration.lst`)
- [ ] I have run the linter (`make -f posix.mak lint`)

File: PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+### Bug fixes:
+
+- [ ] My PR is based of a Bugzilla issue, and I've [referenced it in the commit message](https://github.com/dlang-bots/dlang-bot#automated-references).
+- [ ] I included test(s) for the fixed issue(s)
+- [ ] I targeted `stable` (required for major, critical and regression fixes)
+
+### Core changes:
+
+* [ ] I have written new tests for my changes
```
Comment (JackStouffer) [inline PULL_REQUEST_TEMPLATE.md]: This seems a bit redundant with lines 4 and 10

File: PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+* [ ] I added a [changelog entry](https://github.com/dlang/phobos/tree/master/changelog)
```
Comment (JackStouffer) [inline PULL_REQUEST_TEMPLATE.md]: I would move this to the top section and get rid of this section all together

**Comment by schveiguy** [comment]
@wilzbach I'll start off by saying I like the idea of having a PR template. Thanks for the initiative! But...

I'm not sure I agree with the purpose of your template. It's basically a checklist you fill out to indicate that you did everything correctly. But it really is a pre-flight checklist for submitting PRs, and not relevant to actual submitted PRs. I'd like to see a template of the information reviewers want to see before we can review and pull.

For instance:

```- [ ] My PR is based of a Bugzilla issue, and I've [referenced it in the commit message](https://github.com/dlang-bots/dlang-bot#automated-references).```

Should really be:

```This fixes bugzilla issue #[XXXXX](https://issues.dlang.org/show_bug.cgi?id=XXXXX)```

If you want a pre-flight checklist, I'd have a single line saying:

```Please follow all of the PR guidelines here: (link). Your Pull request will not be pulled until these conditions are met.```

The reason I don't like the checklist is because it's obvious if most of these things are present, and the checkboxes themselves are not necessary for a good PR. What if you include unit tests but didn't check the box? It makes things unnecessarily verbose, and feels like a bureaucratic document.

What I'd have is a blurb at the top that links to the guidelines, then have sections such as "If your PR fixes a bug, use this template. If your PR adds a feature, use this template." Then the user erases the non-relevant templates, and fills out the one that is important.

Things I think could be useful to have in the template:
- Bugzilla issue#
- Relevant forum discussion
- Any dependent PRs
- Any precedent PRs
- If an enhancement that adds to the language, a corresponding spec update PR.
- If an external project is needed to test the veracity of the PR, instructions on how to do so (this would be something that a more seasoned developer would use to add independent unit tests the more novice submitter may not know how to do).

The pre-flight checklist is important too, but obviously would go elsewhere. In that list, I'd like to see a suggestion to tag unittests as `@system` or `@safe`.

Again, thanks for proposing something like this!

**Comment by wilzbach** [comment]
> In that list, I'd like to see a suggestion to tag unittests as @system or @safe.

We already have CircleCi enforcing this 😈

Thinking about this, the better way forward is to add more automation to dlang-bot and our CIs, so that contributors will be automatically warned / guided. Thanks for your input!

**Comment by schveiguy** [comment]
The bot and checks are good, but an actual list of requirements is helpful too.

**Pull request closed without merging**